src/rnnlm/rnnlm-training.cc

```diff
@@ -0,0 +1,587 @@
+    int non_zero_index = -1;                                                    
+    sv.Max(&non_zero_index); 
+    indexes[i].push_back(non_zero_index);
+  }
+
+  vector<vector<BaseFloat> > out;
+
+  output_projection->Propagate(output_0_gpu, indexes, &out);
+//  output_projection->Propagate(*new_output, indexes, &out);
+
+  *tot_weight = post.Sum();
+  *tot_objf = 0;
+  for (int i = 0; i < k; i++) {
+    KALDI_ASSERT(out[i].size() == 1);
+    *tot_objf += log(out[i][0]); // last one (k) is the correct lable
+  }
+
+  if (supply_deriv && nnet != NULL) {
+    // the derivative on the real output
+    vector<vector<BaseFloat> > output_deriv(k);
+
+    for (int i = 0; i < k; i++) {
+      output_deriv[i].push_back(1);
+    }
+
+    // the derivative after the affine layer (before the nonlin)
+
+    // the derivative of the 'nnet3' part
+    CuMatrix<BaseFloat> input_deriv(output_0_gpu.NumRows(),
+                                    output_0_gpu.NumCols(),
+                                    kSetZero);
+
+    CuMatrix<BaseFloat> place_holder;
+    output_projection->Backprop(indexes, output_0_gpu, place_holder,
+                                output_deriv, nnet->output_projection_,
+                                &input_deriv);
+
+//    CuMatrix<BaseFloat> cu_input_deriv(input_deriv);
+
+    computer->AcceptOutputDeriv(output_name, &input_deriv);
+  }
+}
+
+void LmNnetSamplingTrainer::ComputeObjectiveFunctionSample(
+                              const vector<BaseFloat> &unigram,
+                              const GeneralMatrix &supervision,
+                              ObjectiveType objective_type,
+                              const std::string &output_name,
+                              bool supply_deriv,
+                              NnetComputer *computer,
+                              BaseFloat *tot_weight,
+                              BaseFloat *tot_objf,
+                              const LmOutputComponent *output_projection,
+                              Matrix<BaseFloat> *new_output,
+                              LmNnet *nnet) {
+  const CuMatrixBase<BaseFloat> &output_0_gpu = computer->GetOutput(output_name);
+//  Matrix<BaseFloat> output_0(output_0_gpu.NumRows(), output_0_gpu.NumCols());
+//  output_0_gpu.CopyToMat(&output_0);
+  new_output->Resize(output_0_gpu.NumRows(), output_0_gpu.NumCols());
+  output_0_gpu.CopyToMat(new_output);
+  int k = supervision.NumRows();
+
+  KALDI_ASSERT(supervision.Type() == kSparseMatrix);
+  const SparseMatrix<BaseFloat> &post = supervision.GetSparseMatrix();
+
+  vector<int> samples = Select(unigram, k);
+  vector<vector<int> > indexes(k);
+  for (int i = 0; i < k; i++) {
+    indexes[i] = samples;
+
+    const SparseVector<BaseFloat> &sv = post.Row(i);                              
+    int non_zero_index = -1;                                                    
+    sv.Max(&non_zero_index); 
+    indexes[i].push_back(non_zero_index);
```

Comment:
I think you need to think a bit more carefully about what happens when you already sampled one of the output labels.  IMO, what would make more sense here is to first say, "we're going to select all the indexes corresponding to output labels [and use weight 1 for them, because we sampled them w.p. 1], and then sample k items from the remaining unigram distribution [and weight them with 1.0 / [the probability with which we sampled them.]).  And note: the probability with which we sampled them != 1/unigram-prob.  Depends on k, and depends on the things we did not sample because they were labels.


Also, you don't seem to be doing the thing that we discussed, which is to select the same set of output labels for the entire minibatch.  That's the only way to make the matrix multiplication reasonably efficient on GPU.